Mrtk wireframe shader fix for Android

Alexees opened the pull request:
## Overview
On Android, opening the shader shows this warning;
`Shader error in 'Mixed Reality Toolkit/Wireframe': variable 'o' used without having been completely initialized at line 121 (on gles3)`

The similarity in names of UNITY_INITIALIZE_VERTEX_OUTPUT_STEREO with UNITY_TRANSFER_VERTEX_OUTPUT_STEREO near that error and the corresponding struct definition led me to this page:

https://forum.unity.com/threads/unity-stereo-output-function.470331/

and I just tried adding UNITY_INITIALIZE_VERTEX_OUTPUT_STEREO before the UNITY_TRANSFER_VERTEX_OUTPUT_STEREO cmd and the error was gone.

No idea if that is correct though

Fixes part of: #5844

wiwei commented on Assets/MixedRealityToolkit/StandardAssets/Shaders/MRTK_Wireframe.shader:
```diff
@@ -115,7 +115,8 @@ Shader "Mixed Reality Toolkit/Wireframe"
                    o.inverseW = 1.0 / o.viewPos.w;
                    o.dist = distScale[idx] * o.viewPos.w * wireScale;
-                   UNITY_TRANSFER_VERTEX_OUTPUT_STEREO(i[idx], o);
+				   UNITY_INITIALIZE_VERTEX_OUTPUT_STEREO(o);
```
```suggestion
                   UNITY_INITIALIZE_VERTEX_OUTPUT_STEREO(o);**
```

keveleigh replied:
@wiwei Intentional `**` at the end of the line in your suggestion?

wiwei replied:
Oh no, not intentional there. Just wanted to get rid of the tabs

wiwei commented on Assets/MixedRealityToolkit/StandardAssets/Shaders/MRTK_Wireframe.shader:
```diff
@@ -115,7 +115,8 @@ Shader "Mixed Reality Toolkit/Wireframe"
                    o.dist = distScale[idx] * o.viewPos.w * wireScale;
-                   UNITY_TRANSFER_VERTEX_OUTPUT_STEREO(i[idx], o);
+				   UNITY_INITIALIZE_VERTEX_OUTPUT_STEREO(o);
+				   UNITY_TRANSFER_VERTEX_OUTPUT_STEREO(i[idx], o);
```
```suggestion
                   UNITY_TRANSFER_VERTEX_OUTPUT_STEREO(i[idx], o);
```

Cameron-Micka replied:
I would recommend moving UNITY_INITIALIZE_VERTEX_OUTPUT_STEREO to just below line 110 (g2f o;) so that the vertex get's initialized just once (and not every loop). I'm still not clear why we would need to initialize the 'o' since it looks like are members are getting set? If you look at this [built in Unity geometry shader](https://github.com/TwoTailsGames/Unity-Built-in-Shaders/blob/master/DefaultResourcesExtra/VR/Shaders/SpatialMappingWireframe.shader) we could try follow that macro placement to see if the error presists on Android.

Cameron-Micka replied:
Maybe this is the macro we need: UNITY_SETUP_STEREO_EYE_INDEX_POST_VERTEX https://docs.unity3d.com/Manual//Android-SinglePassStereoRendering.html I can't find much documentation on it, but it is called out in that android specific doc.

SimonDarksideJ commented:
Going to sleep now, but I tried the shader in a clean project (just taking the shader file) and tested on Quest. No joy, still does not render.

FYI, we already had the same Initialize function in the XRTK wireframe shader previously and it wasn't working.

Alexees commented:
@SimonDarksideJ

This is my setup:
Unity 2018.4.8f1
I added the Oculus Virtual Reality SDK
Minimum API level 19
Changed package name
Used a single cube in the scene with the shader on it

This is the current Oculus Quest software status:
![image](https://user-images.githubusercontent.com/32508825/65222131-ba8db000-dabe-11e9-8201-591e19ff2c0e.png)

Android:
SDK Installed via Android Studio 10.0 (Q)
Embedded JDK 1.8.0_152

This is my project that successfully builds
[OculusQuestGeometry.zip](https://github.com/microsoft/MixedRealityToolkit-Unity/files/3629942/OculusQuestGeometry.zip)

SimonDarksideJ commented:
So I have good news and bad news.

Good news, you are indeed correct that the updated shader does in fact run on the Quest in a clean project, even with MRTK installed in the project.

However.

If you add the MRTK to the scene to control the camera, it fails utterly.  Tested in XRTK and it's exactly the same response.

As a side note, once I added the MRTK to the scene, I also got a load of other shader compilation errors for the standard shader.  Odd they should only appear once MRTK was "active"

Alexees commented:
@SimonDarksideJ well... it's no geom then afterall. Do you know how make the Quest choose a different shader pass? How this that even picked as a suitable one? I mean it should just pick the Diffuse Fallback shader, right?

SimonDarksideJ commented:
I think you miss-understand me @Alexees , I see this as positive news.  it proves that geom shaders to in fact work (despite Oculus noting they don't?) and that the problem as to why it's drawing darkly on the Quest is an issue with the MRTK camera profile.

I've just confirmed that by disabling the camera profile and confirmed it makes it operational again.

So the next task is to find what is causing the issue and find a resolution.

Likely the shader is in fact rendering but due to some camera setting is causing it to draw black, hence why it is not falling back to diffuse.

The mystery continues.

Alexees commented:
Ah, nice... I thought an "in isolation working shader" is no good, especially with the errors provided in logcat. Good to hear that there is hope :D

SimonDarksideJ commented:
Resolution found, an in hindsight a bloomin obvious one.

I'll simply put this screenshot up and see if you can guess which specific setting is causing the issue on the Quest (mobile) platform.

![image](https://user-images.githubusercontent.com/1793042/65247649-ad85b680-dae8-11e9-84be-98fc15bf3bb7.png)

Answers on a postcard, winner gets the kudos of figuring it out second.

Alexees commented:
Neat hint (mobile)... It's probably the ultra settings then...

SimonDarksideJ commented:
Give that man a prize,  set it to Medium and work world is in wireframe again.
You still need that PR to add the initialise call for it to work.

Although, I suggest further testing as I had to remove the standard shader as it caused issues.

Alexees commented:
> Give that man a prize, set it to Medium and work world is in wireframe again.
> You still need that PR to add the initialise call for it to work.
> 
> Although, I suggest further testing as I had to remove the standard shader as it caused issues.

I hope @Cameron-Micka is working on that. I gave him a pretty stripped version of the shader part that doesn't work on Android:

https://github.com/microsoft/MixedRealityToolkit-Unity/issues/5844#issuecomment-532183334

Alexees commented:
> set it to Medium and work world is in wireframe again

To me this looks like a place where refactoring is an option. Providing per Platform settings, for at least the quality, although I have no idea what settings are mandatory anyways...

Cameron-Micka commented:
Great find @SimonDarksideJ! I’ve always thought opaque displays should not default to Ultra (we are trying to hit 90fps with stereo rendering so that’s not a lot of milliseconds to play with). I’m going to take a look at the other MRTK shader errors on Android next week.

SimonDarksideJ commented:
Note @Cameron-Micka , this still needs the initialise call, else it fails compilation on Android.

Alexees commented:
That was a little too quick in cleaning PRs...
